### PR TITLE
Add settings to rustdoc to use the system theme

### PR DESCRIPTION
**Pull request by nasso**
This PR adds new settings to `rustdoc` to use the operating system color scheme.

click

`rustdoc` actually had [basic support for this](https://github.com/rust-lang/rust/blob/b1af43bc63bc7417938df056f7f25d456cc11b0e/src/librustdoc/html/static/storage.js#L121), but the setting wasn't visible and couldn't be set back once the theme was explicitly set by the user. It also didn't update if the operating system theme preference changed while viewing a page.

I'm using [this method](https://developer.mozilla.org/en-US/docs/Web/CSS/Media_Queries/Testing_media_queries#Receiving_query_notifications) to query and listen to changes to the `(prefers-color-scheme: dark)` media query. I kept the old method (based on `getComputedStyle`) as a fallback in case the user-agent doesn't support `window.matchMedia` (so like... [pretty much nobody](https://caniuse.com/?search=matchMedia)).

Since there's now more than one official ""dark"" theme in `rustdoc` (and also to support custom/third-party themes), the preferred dark and light themes can be configured in the settings page (the defaults are just "dark" and "light").

This is also my very first "proper" PR to Rust! Please let me know if I did anything wrong :).

### CODE REVIEW

File: src/librustdoc/html/render/mod.rs [hunk 1]
```diff
@@ -1370,13 +1385,49 @@ impl Setting {
...
+}
+
+impl From<(&'static str, &'static str, &'static str, Vec<(String, String)>)> for Setting {
+    fn from(values: (&'static str, &'static str, &'static str, Vec<(String, String)>)) -> Setting {
+        Setting::Select {
+            js_data_name: values.0,
+            description: values.1,
+            default_value: values.2,
+            options: values.3,
+        }
```
Comment (jyn514) [inline src/librustdoc/html/render/mod.rs:1432]: This looks pretty confusing, I would just construct it explicitly with struct initialization syntax.

**Review by jyn514 — COMMENTED** [review]
Small nit, but I think guillaume is the better reviewer for this. This looks really cool, thanks for working on it!

r? @GuillaumeGomez

File: src/librustdoc/html/render/mod.rs (hunk 1 above)
Comment (nasso) [reply src/librustdoc/html/render/mod.rs:1432]: Good point, I tried to follow what was already here but since it was just on/off toggles it wasn't as complex. I'll change that tomorrow!

File: src/librustdoc/html/render/mod.rs [hunk 2]
```diff
@@ -1343,22 +1345,35 @@ impl AllTypes {
...
                 "<div class='setting-line'>\
                      <div class='title'>{}</div>\
                      <div class='sub-settings'>{}</div>
                  </div>",
                 description,
-                sub_settings.iter().map(|s| s.display()).collect::<String>()
+                sub_settings.iter().map(|s| s.display(root_path, suffix)).collect::<String>()
             ),
-            Setting::Entry { ref js_data_name, ref description, ref default_value } => format!(
+            Setting::Toggle { ref js_data_name, ref description, ref default_value } => format!(
```
Comment (pickfire) [inline src/librustdoc/html/render/mod.rs:1376]: I wonder if we still need all these `ref`?

File: src/librustdoc/html/render/mod.rs [hunk 3]
```diff
@@ -1370,13 +1385,40 @@ impl Setting {
                 if *default_value { " checked" } else { "" },
                 description,
             ),
+            Setting::Select {
+                ref js_data_name,
+                ref description,
+                ref default_value,
+                ref options,
+            } => format!(
+                "<div class='setting-line'>\
```
Comment (pickfire) [inline src/librustdoc/html/render/mod.rs:1394]: I think you might want to use `"` rather than `'`, I recently just send a patch on that.

File: src/librustdoc/html/static/settings.css [hunk 4]
```diff
@@ -30,6 +29,45 @@
 	display: none;
 }
 
+.select-wrapper {
+	float: right;
+
+	position: relative;
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/settings.css:35]: I'm just wondering: is there any particular reason for these empty lines?

File: src/librustdoc/html/static/settings.js [hunk 5]
```diff
@@ -2,29 +2,54 @@
...
         }
-        for (var i = 0; i < elems.length; ++i) {
-            var toggle = elems[i].previousElementSibling;
-            var settingId = toggle.id;
-            var settingValue = getSettingValue(settingId);
-            if (settingValue !== null) {
-                toggle.checked = settingValue === "true";
+
+        if (elems.selects && elems.selects.length > 0) {
+            for (var i = 0; i < elems.selects.length; ++i) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/settings.js:42]: `i` has already been declared in the previous loop (in case the previous loop was run), so please create the `i` variable on its own before the two loops (`var i;`).

File: src/librustdoc/html/static/settings.css [hunk 6]
```diff
@@ -4,7 +4,6 @@
 }
 
 .setting-line > div {
-	max-width: calc(100% - 74px);
```
Comment (pickfire) [inline src/librustdoc/html/static/settings.css:7]: Why is this removed?

File: src/librustdoc/html/static/settings.css (hunk 4 above)
Comment (nasso) [reply src/librustdoc/html/static/settings.css:35]: Oh my bad, that's just a habit I have. I tend to try and separate groups of rules that kind of go together (e.g. position followed by top/right/left/bottom). I'll remove them 👍

File: src/librustdoc/html/static/settings.css (hunk 6 above)
Comment (nasso) [reply src/librustdoc/html/static/settings.css:7]: With the `<select>` element being aligned to the right, this rule made it awkwardly stand 74 pixels away from the end of the page.

Before:
image

After:
image


It didn't really make sense to me so I removed it. Note that it wasn't really useful before because all the content was aligned to the left anyway, so I don't think it broke anything else.

File: src/librustdoc/html/render/mod.rs (hunk 2 above)
Comment (nasso) [reply src/librustdoc/html/render/mod.rs:1376]: Nope, not for the `&'static str`s and `bool`s. Only needed for types that aren't `Copy`. I'm removing them!

File: src/librustdoc/html/render/mod.rs (hunk 3 above)
Comment (nasso) [reply src/librustdoc/html/render/mod.rs:1389]: Why? As far as I know, they're equivalent and don't require escaping here.

File: src/librustdoc/html/static/settings.js (hunk 5 above)
Comment (nasso) [reply src/librustdoc/html/static/settings.js:42]: The variables are scoped to the loop anyway. I created the variable on its own like you asked but may I know why it's preferred?

Comment (jyn514) [reply src/librustdoc/html/static/settings.js:42]: I don't think they *are* scoped to the loop - javascript has variable hoisting.

Comment (jyn514) [reply src/librustdoc/html/static/settings.js]:
```
(bash@joshua-ThinkPad-P51s) ~ ▶️ node
Welcome to Node.js v12.16.1.
Type ".help" for more information.
>  for (var i = 0; i < 5; ++i) {}
undefined
> i
5
```

Comment (nasso) [reply src/librustdoc/html/static/settings.js]: Oh my god. My bad. I got used to using `let` and forgot the semantics of plain-old `var`iables...

**Comment by nasso** [comment]
I'm not sure I understand how the current implementation of the theme system works. [It says here](https://github.com/rust-lang/rust/blob/bc74dd711fd154555dea3b110dfed39c4dc37bc6/src/librustdoc/html/render/mod.rs#L476) that it works by changing the href of the `light.css` `<link>` tag. But then, what's the point of including `<link>` tags for every other theme (especially if they stay `disabled`).

I'd like to know so I can properly implement a fix to support JavaScript disabled browsers to use the system theme :)

File: src/librustdoc/html/static/settings.js (hunk 5 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/settings.js]: The "surprises" of javascript. Don't worry, we all went through this. ;)

**Comment by jyn514** [comment]
@Cldfire might know ^

**Comment by GuillaumeGomez** [comment]
The `disabled` attribute is used to prevent the style to be applied, however they still are loaded by the browser in case you want to switch theme. Which means that when you arrive on a page, you won't have the light theme then the selected theme displayed (it would look like a weird flash haha).

**Comment by Cldfire** [comment]
I tried getting rid of the `disabled` attribute in https://github.com/rust-lang/rust/pull/75063, but unfortunately I wasn't able to find a way to get rid of it in a way that's supported everywhere and doesn't result in content flashing on page load.

Usage of the `disabled` attribute was first introduced in https://github.com/rust-lang/rust/pull/71237; prior to that PR all themes were enabled at the same time and therefore had to have exactly the same CSS rules (otherwise they'd conflict with one another).

**Comment by Nemo157** [comment]
What is the forward/backwards compatibility of this? All docs hosted on docs.rs are using a shared `localStorage` but the javascript from when they were built, so will this behave well when you change the settings in docs with this new JS then open docs with the old JS?

**Comment by nasso** [comment]
I didn't change the way `localStorage` is used at all. I only added new settings, and the default theme behavior (system theme) will be kept. So it's 100% backwards compatible!

**Comment by Nemo157** [comment]
(This also looks relatively easy for docs.rs to hijack to configure its own theme (when we get support), other than the lack of self-`storage`-events to make it easy for us to detect when to check the value).

File: src/librustdoc/html/render/mod.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:1389]: It's about unification. Nothing more. Please do so then. ;)

**Comment by GuillaumeGomez** [comment]
@Nemo157 Yes, I said a while ago that I needed to do it but still haven't found time to do so... I'll try to get back to it once this PR is merged.

@nasso Looks all good to me, thanks! Just remains to replace single quotes with double quotes and then it's good to go.

File: src/librustdoc/html/render/mod.rs (hunk 3 above)
Comment (nasso) [reply src/librustdoc/html/render/mod.rs:1389]: Oh alright! Should I also change all the other simple quotes that were already used in the code?

Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:1389]: iirc, @pickfire has a PR ready for them, so just do it for the ones you added please.

**Comment by nasso** [comment]
Actually I think I should make sure it works properly with JavaScript disabled before merging. Currently it doesn't work, because the logic is in `storage.js`. One way i see to fix it would be to add a `default-theme.css` stylesheet that would look like this:

```css
@import "{static_root_path}light{suffix}.css" (prefers-color-scheme: light);
@import "{static_root_path}dark{suffix}.css" (prefers-color-scheme: dark);
```

The problem with this is that the styles might override each other in bad ways (since there's already the javascript controlled `<link>`). My idea was to make that `<link>` default to `default-theme.css` instead of `light.css`, and let JavaScript override it if necessary. I don't know if this will cause content flashing though, I'll have to test it first if you don't see any other issue it might cause.

**Comment by GuillaumeGomez** [comment]
> Actually I think I should make sure it works properly with JavaScript disabled before merging.

By default it's the light theme that gets displayed, so if you get the same result, it's working the same as currently! But if you want to extend it (which would be nice!), please do so but in another PR so we can get this one merged.

**Comment by nasso** [comment]
Actually the old behavior was using the system theme by default! You can try it by opening [the standard library documentation](https://doc.rust-lang.org/std/index.html) in a private window (so that `localStorage` is empty). According to your system's theme, it will either use the "dark" or the "light" theme. Only thing is that you have to reload the page to update the theme when you change it in your operating system settings.

**Comment by GuillaumeGomez** [comment]
Hmmmm... Makes sense in a way I guess?

**Comment by nasso** [comment]
Yeah, because the old method used [`getSystemTheme`](https://github.com/rust-lang/rust/blob/d9b931669b9c480b60f0abcaa7834758ba8761a2/src/librustdoc/html/static/storage.js#L121) as the default value if `rustdoc-theme` wasn't set in `localStorage`. The "light" theme was only the second fallback. See [here in `storage.js`](https://github.com/rust-lang/rust/blob/d9b931669b9c480b60f0abcaa7834758ba8761a2/src/librustdoc/html/static/storage.js#L127).

**Comment by jyn514** [comment]
> According to your system's theme, it will either use the "dark" or the "light" theme. 

Isn't that the intended behavior? What more is there to change?

**Comment by nasso** [comment]
> 
> 
> > According to your system's theme, it will either use the "dark" or the "light" theme.
> 
> Isn't that the intended behavior? What more is there to change?

Aah nevermind, I thought I was changing the behavior, but actually even the old implementation didn't use the system theme if JavaScript was disabled. I'll get this PR ready to be merged then 👍

**Comment by nasso** [comment]
It should be ready for merge now 👍

File: src/librustdoc/html/static/storage.js [hunk 7]
```diff
@@ -110,19 +110,79 @@ function switchTheme(styleElem, mainStyleElem, newTheme, saveTheme) {
...
+    };
+})();
+
+if (getCurrentValue("rustdoc-use-system-theme") !== "false" && window.matchMedia) {
+    // call the function to initialize the theme at least once!
+    updateSystemTheme();
+} else {
+    switchTheme(currentTheme, mainTheme,
+                getCurrentValue("rustdoc-theme") || "light",
+                false);
```
Comment (pickfire) [inline src/librustdoc/html/static/storage.js:187]:
```suggestion
    switchTheme(
        currentTheme,
        mainTheme,
        getCurrentValue("rustdoc-theme") || "light",
        false
    );
```
I don't think we need to make this indentation feels special from the rest.

Comment (nasso) [reply src/librustdoc/html/static/storage.js]: Correct, I wanted to keep it like it was before but it makes more sense like this!

**Review by pickfire — APPROVED** [review]
Looks good to me.

**Comment by jyn514** [comment]
@nasso do you mind squashing the commits a little? There are instructions at https://rustc-dev-guide.rust-lang.org/git.html#advanced-rebasing.

File: src/librustdoc/html/static/storage.js (hunk 7 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/storage.js]: :+1:

**Comment by nasso** [comment]
I hope I did it right!

**Comment by GuillaumeGomez** [comment]
Looks good to me! r=me and @jyn514 once CI pass.

**Comment by nasso** [comment]
One thing to note though: users who have previously changed their theme to `dark` or `ayu` might have it overridden by the new default settings (which uses the system theme based on the `rustdoc-use-system-theme` variable in `localStorage`).

Example for someone using a "dark" system theme:
- Before this PR, they manually changed Rustdoc theme to the new `"ayu"` theme
- This changed `localStorage["rustdoc-theme"]` to `"ayu"`.
- After the PR, Rustdoc checks for `localStorage["rustdoc-use-system-theme"]`, which isn't defined. It defaults to `"true"`.
- Rustdoc detects that the preferred color scheme is dark, so it switches the theme to `localStorage["rustdoc-preferred-dark-theme"]`. This isn't defined either, so it uses the default value: `"dark"`.
- The user is annoyed because Rustdoc changed their theme! They don't know about the new theme settings, so they just manually change the theme back to `"ayu"`.
- When you manually change the theme, Rustdoc automatically turns off the "Use system theme" option.
- Rustdoc no longer updates with the system theme (if it ever changes, which it might do at sunset!).

The user never knew Rustdoc could do that in the first place, and only had their theme randomly changed to `"dark"` once and never had to worry about it ever again. They're keep manually changing the theme like they're used to.

Thoughts on this?

**Comment by jyn514** [comment]
Also, if `rustdoc-theme` is set to a dark theme but the system theme is a light theme, can you set `use-system-theme` to false?

**Comment by jyn514** [comment]
@bors r-

I'd like to address the issues @nasso  brought up (thanks for thinking of them!)

**Comment by nasso** [comment]
I can implement that logic! I suppose that for now I should hard-code what theme is "dark" and what theme is "light"?
At least until #74937 is fixed.

**Comment by nasso** [comment]
> Also, if `rustdoc-theme` is set to a dark theme but the system theme is a light theme, can you set `use-system-theme` to false?

This can cause issues:

1. Someone has their operating system switch automatically between dark mode and light mode according to the time of the day.
2. Rustdoc is all using default settings (assuming an empty `localStorage`). That means it follows the system theme, and writes it to `localStorage` when switching to it.
3. At night, Rustdoc is using the default `dark` theme. This is just the default behavior. Now `localStorage["rustdoc-theme"] === "dark"`.
4. The next day, `localStorage["rustdoc-theme"] === "dark"`, but the OS is now in light mode (it's daytime).
5. Rustdoc sets `localStorage["rustdoc-use-system-theme"]` to `false`, thinking it was a user of the older Rustdoc version!

One potential fix would be to not save the theme to `localStorage` when automatically switched to. The downside is that older versions of Rustdoc won't pick up the theme because it's not stored there anymore.

**Comment by jyn514** [comment]
> Someone has their operating system switch automatically between dark mode and light mode according to the time of the day.

Seems a bit of an edge case ... but since it leads to rustdoc overriding the user's theme, better to avoid it. Rather than having complicated logic, let's just only implement my first suggestion: If use-system-theme and preferred-dark-theme aren't set, but rustdoc-theme is set to a dark theme, update preferred-dark-theme to that theme.

**Comment by pickfire** [comment]
> Someone has their operating system switch automatically between dark mode and light mode according to the time of the day.

Is it an edge case? My phone does this too.

**Comment by jyn514** [comment]
> I suppose that for now I should hard-code what theme is "dark" and what theme is "light"?

You can write the canonical list :) There's only three themes, it seems reasonable to have a `match` on them somewhere.

**Comment by nasso** [comment]
> If use-system-theme and preferred-dark-theme aren't set, but rustdoc-theme is set to a dark theme, update preferred-dark-theme to that theme.

That should be implemented now! I think it's ready now.

**Comment by jyn514** [comment]
Looks great! Thanks so much for working on this :)

@bors r=jyn514,guillaumegomez

**Comment by pickfire** [comment]
I wonder if we can have a css fallback rather than relying on javascript, some users may have disabled javascript.

**Comment by nasso** [comment]
> I wonder if we can have a css fallback rather than relying on javascript? Some users may have disabled javascript.

I thought about this already and I think I may implement a CSS fallback in a future PR. I already have a few ideas about how it could work but I'd prefer to discuss about it first :)
